File: scripts/fetch-exercisedb.mjs
import fs from 'fs';

const BASE = 'https://oss.exercisedb.dev/api/v1/exercises';
const OUT = '/Users/ignaziofiorito/benessere-forma/scripts/exercisedb-catalog.json';
const DELAY_MS = 500;
const MAX_RETRIES = 6;

const sleep = (ms) => new Promise(r => setTimeout(r, ms));

async function fetchPage(after) {
  const params = new URLSearchParams({ limit: '25' });
  if (after) params.set('after', after);
  let lastErr;
  for (let attempt = 0; attempt < MAX_RETRIES; attempt++) {
    if (attempt > 0) {
      const backoff = Math.min(15000, 2000 * Math.pow(2, attempt - 1));
      console.log(`  retry ${attempt} in ${backoff}ms...`);
      await sleep(backoff);
    }
    const resp = await fetch(`${BASE}?${params.toString()}`);
    if (resp.ok) return resp.json();
    lastErr = `HTTP ${resp.status}`;
    if (resp.status !== 429 && resp.status < 500) {
      throw new Error(`${lastErr}: ${(await resp.text()).slice(0,200)}`);
    }
  }
  throw new Error(`Exhausted retries: ${lastErr}`);
}

const all = [];
const seen = new Set();
let after = null;
let page = 0;
const t0 = Date.now();

while (true) {
  page++;
  const json = await fetchPage(after);
  const data = json.data || [];
  let newCount = 0;
  for (const e of data) {
    const id = e.exerciseId || e.id;
    if (!seen.has(id)) { seen.add(id); all.push(e); newCount++; }
  }
  const meta = json.meta || {};
  console.log(`Page ${page}: +${newCount} new (running ${all.length} / target ${meta.total ?? '?'})`);

  fs.writeFileSync(OUT, JSON.stringify(all, null, 2));

  if (!meta.hasNextPage || !meta.nextCursor || newCount === 0) break;
  after = meta.nextCursor;
  await sleep(DELAY_MS);
}

console.log(`\nDone in ${((Date.now()-t0)/1000).toFixed(1)}s. Total unique: ${all.length}`);
console.log(`Saved -> ${OUT}`);

File: scripts/fetch-exercisedb.py
import json
import time
import urllib.request
import urllib.parse

BASE = "https://oss.exercisedb.dev/api/v1/exercises"
OUT = "/Users/ignaziofiorito/benessere-forma/scripts/exercisedb-catalog.json"

all_exercises = []
cursor = None
page = 0
t0 = time.time()

while True:
    params = {"limit": 100}
    if cursor:
        params["cursor"] = cursor
    url = f"{BASE}?{urllib.parse.urlencode(params)}"
    with urllib.request.urlopen(url, timeout=20) as resp:
        body = json.loads(resp.read())

    page += 1
    chunk = body.get("data", [])
    all_exercises.extend(chunk)
    meta = body.get("meta", {})
    cursor = meta.get("nextCursor")
    has_next = meta.get("hasNextPage", False)
    print(f"Page {page}: +{len(chunk)} (total so far: {len(all_exercises)} / {meta.get('total','?')})")
    if not has_next or not cursor:
        break

print(f"\nDone in {time.time()-t0:.1f}s. Total: {len(all_exercises)} exercises")

with open(OUT, "w") as f:
    json.dump(all_exercises, f, indent=2)

print(f"Saved to {OUT}")
